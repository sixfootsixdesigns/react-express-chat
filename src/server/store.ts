import * as shortid from 'shortid';

export interface IMessage {
  author: IUser;
  message: string;
  id: string;
}

export interface IUser {
  name: string;
  id: string;
}

const MAX_MESSAGES = 50;
let users: IUser[] = [];
let messages: IMessage[] = [];

const addMessage = (author: IUser, message: string) => {
  const data: IMessage = {
    author: author,
    message: message,
    id: shortid.generate()
  };
  messages.push(data);

  messages = messages.slice(MAX_MESSAGES * -1);

  return data;
};

const dropUser = (remove: IUser) => {
  users = users.filter(user => user.id !== remove.id);
};

const addUser = (username: string) => {
  const user: IUser = {
    name: username,
    id: shortid.generate()
  };
  users.push(user);
  return user;
};

const getUsers = () => users;
const getMessages = () => messages;

const reset = () => {
  users = [];
  messages = [];
};

export const storeAPI = {
  MAX_MESSAGES,
  addUser,
  getUsers,
  getMessages,
  addMessage,
  reset,
  dropUser
};
