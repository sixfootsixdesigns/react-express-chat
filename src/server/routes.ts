import { Router } from 'express';
import { storeAPI } from './store';
import socket from 'socket.io';

export const routes = (socket: socket.Server) => {
  const router = Router();

  router
    .get('/messages', (req: Request, res: any) => {
      res.json(storeAPI.getMessages());
    })

    .post('/add-message', (req: Request, res: any) => {
      const body: any = req.body;
      const message = storeAPI.addMessage(body.author, body.message);

      socket.emit('NewMessage', storeAPI.getMessages());

      res.status(201);
      res.json(message);
    })

    .get('/users', (req: Request, res: any) => {
      res.json(storeAPI.getUsers());
    })

    .post('/add-user', (req: Request, res: any) => {
      const body: any = req.body;
      const user = storeAPI.addUser(body.username);

      socket.emit('UserAdded', storeAPI.getUsers());

      res.status(201);
      res.json(user);
    });

  return router;
};
