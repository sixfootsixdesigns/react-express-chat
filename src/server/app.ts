import express from 'express';
import * as bodyParser from 'body-parser';
import { routes } from './routes';
import * as http from 'http';
import socket from 'socket.io';
import { storeAPI, IUser } from './store';

const app = express();

const appServer = http.createServer(app);
const socketIO = socket(appServer);

app
  .use(bodyParser.json())
  .use(routes(socketIO))
  .use(express.static('build'));

socketIO.on('connection', socket => {
  socket.on('disconnecting', (user: IUser) => {
    storeAPI.dropUser(user);
    socketIO.emit('UserLeft', storeAPI.getUsers());
  });
});

export { appServer, socketIO };
