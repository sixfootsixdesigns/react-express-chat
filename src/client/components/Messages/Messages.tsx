import React from 'react';
import { IMessage } from '../../../server/store';

interface MessagesProps {
  messages: IMessage[];
  loading?: boolean;
}

export class Messages extends React.Component<MessagesProps> {
  public render() {
    const { loading, messages } = this.props;

    if (loading) {
      return <div className="Messages">loading...</div>;
    }

    return (
      <div className="Messages">
        {(messages.length &&
          messages.map(message => (
            <div key={message.id}>
              <strong>{message.author.name}</strong>: {message.message}
            </div>
          ))) || <div>No Messages Yet.</div>}
      </div>
    );
  }
}
