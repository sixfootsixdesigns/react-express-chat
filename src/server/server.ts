import { appServer } from './app';

const port = Number(process.env.PORT || 4001);

appServer.listen(port, () => {
  console.log(`listening on port: ${port}`);
});
