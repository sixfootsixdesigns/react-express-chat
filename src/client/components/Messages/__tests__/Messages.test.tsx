import * as React from 'react';
import { render, cleanup } from 'react-testing-library';
import { Messages } from '../Messages';
import { IMessage } from '../../../../server/store';

describe('Messages Component', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render with messages', () => {
    const messages: IMessage[] = [
      {
        author: {
          name: 'bob',
          id: 'abc'
        },
        message: 'hi',
        id: '1a'
      },
      {
        author: {
          name: 'dave',
          id: '34ax'
        },
        message: 'hello',
        id: 'zz3'
      }
    ];

    const { container } = render(<Messages messages={messages} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without messages', () => {
    const messages: IMessage[] = [];

    const { container } = render(<Messages messages={messages} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render loading state', () => {
    const messages: IMessage[] = [];

    const { container } = render(
      <Messages loading={true} messages={messages} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
