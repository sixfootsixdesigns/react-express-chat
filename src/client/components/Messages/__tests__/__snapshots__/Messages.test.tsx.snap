// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages Component should render loading state 1`] = `
<div
  class="Messages"
>
  loading...
</div>
`;

exports[`Messages Component should render with messages 1`] = `
<div
  class="Messages"
>
  <div>
    <strong>
      bob
    </strong>
    : 
    hi
  </div>
  <div>
    <strong>
      dave
    </strong>
    : 
    hello
  </div>
</div>
`;

exports[`Messages Component should render without messages 1`] = `
<div
  class="Messages"
>
  <div>
    No Messages Yet.
  </div>
</div>
`;
